trace_events: add traced_value.cc/traced_value.h

jasnell opened the pull request:
Port of the V8 internal v8::tracing::TracedValue that allows
structured data to be included in the trace event. The v8 class
is not exported in the public API so we cannot use it directly.

This is a simplified and slightly modified port. This commit only
adds the class, it does not add uses of it. Those will come in
separate PRs/commits.

This *will* be used to include more complex data structures within
various trace events.

/cc @ofrobots @eugeneo @nodejs/diagnostics @nodejs/trace-events 

##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/15567/

jasnell commented:
ping @nodejs/diagnostics

mike-kaufman commented on src/tracing/traced_value.h:
```diff
@@ -0,0 +1,68 @@
+  void WriteName(const char* name);
+
+  std::string data_;
+  bool first_item_;
```
not following how this is being used, and how its value is correct after a call sequence like `BeginArray`, `BeginDictionary`, `EndDictionary`.

jasnell replied:
Look within the `WriteComma` function for use (https://github.com/nodejs/node/pull/21475/files#diff-7e8d44b3c53ea475e1b34c59578ed6e8R169) ... 

The original implementation in V8 has `DCHECK` statements that are only enabled in debug builds that check for proper state. I pulled those to keep things simple but could add them back in as regular `CHECK`s

addaleax commented on src/tracing/traced_value.cc:
```diff
@@ -0,0 +1,191 @@
+
+namespace {
+
+void EscapeAndAppendString(const char* value, std::string* result) {
```
nit: It would be nicer to return a `std::string` instead of making `result` an in/out parameter, tbh…

jasnell replied:
I agree. I adopted this directly from the v8 impl. If it's ok to diverge then I can change

addaleax commented on src/tracing/traced_value.cc:
```diff
@@ -0,0 +1,191 @@
+      std::ostringstream stream;
+      stream.imbue(std::locale("C"));  // Ignore locale
+      stream << v;
+      return stream.str();
```
Does `return std::to_string(v);` do the trick here too, for the `default` case?

jasnell replied:
I'll double check

jasnell replied:
Ah, right, the challenge with `std::to_string(v)` is that it ignores any reasonable precision on the double and always prints six decimals, whereas the version used here will print with actual scientific notation (e.g. `std::to_string(v)` when `v = 1.23e7`, will print `12300000.000000` whereas the version used here will print `1.23e+07` as one would more reasonably expect.

addaleax commented on src/tracing/traced_value.cc:
```diff
@@ -0,0 +1,191 @@
+void TracedValue::AppendAsTraceFormat(std::string* out) const {
+  *out += root_is_array_ ? '[' : '{';
+  *out += data_;
+  *out += root_is_array_ ? ']' : '}';
```
Same here, seems like it might be nicer to make `out` a return value rather than an in/out parameter

jasnell replied:
This is part of the v8 trace event API and really isn't under our control, unfortunately.

jasnell commented:
@mcollina @addaleax ... PTAL, I've updated the escaping to properly handle UTF8 characters, added the requested escaping test, and made a couple other minor improvements.

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/15817/

jasnell commented:
@addaleax ... to give you a better idea about how this is going to be used... here's an example

```c
auto traced_value = TracedValue::Create();
traced_value->SetString("abc", "xyz");
TRACE_EVENT_NESTABLE_ASYNC_BEGIN1("cat", "name", 1, "data", traced_value);
```

Within the trace event log, the `traced_value` will be serialized as additional JSON payload as the value of the `data` argument. What this allows us to do is provide complex structured and nested data within the trace event as opposed to simple primitives. One example of this will be trace events for http and http2, which will optionally include detailed information about the session, the headers, etc.

jasnell commented:
Related failures in Windows and a couple others... trying again: https://ci.nodejs.org/job/node-test-pull-request/15819/

Trott commented:
Unfortunately, for windows-fanned jobs, "Resume Build" does not work. (I don't know if this is something Build WG can fix or not.) So for that, you need to use Rebuild on the windows-fanned job itself instead.

Regardless, it looks like you've pushed two commits since the last full CI run (if GitHub's interface isn't misleading me somehow) so a full CI run is in order.

CI: https://ci.nodejs.org/job/node-test-pull-request/15821/

Trott commented:
Hmm...lots of stuff failing very fast. Something odd is up...

Trott commented:
Ah, the problem might be test-softlayer-ubuntu1604-x64-1 which builds for a few different tasks. I'll see if someone is around on #node-build who might know how to investigate/fix.

jasnell commented:
@Trott ... I was not using "Resume Build". When adding new commits, I always start a full CI run.

jasnell commented:
New CI Run: https://ci.nodejs.org/job/node-test-pull-request/15823/

jasnell commented:
woo... green CI.

jasnell commented:
Landed in d85449d
